src/support/hash.h

```diff
@@ -22,27 +22,22 @@
 
 namespace wasm {
 
-typedef uint32_t HashType;
```

Comment:
1. It may still be nice to have a typedef of HashType, instead of hardcoding `size_t` in various places? I don't feel strongly though.
2. Does this change from 32 to 64 bit hashes have a performance impact?

Reply:
> It may still be nice to have a typedef of HashType, instead of hardcoding size_t

Shouldn't be a problem, though building upon `std::hash` does effectively pin the type to `size_t` anyway, i.e. what the standard library uses and documents. For [example](https://en.cppreference.com/w/cpp/utility/hash):

> **std::hash**
> 2. Returns a value of type `std::size_t` that represents the hash value of the parameter.

> Does this change from 32 to 64 bit hashes have a performance impact?

I'd expect the difference to be negligible. Looking at the VS implementation as an example, it boils down to doing 64-bit instead of 32-bit `XOR` and `MUL` on 64-bit platforms:

```cpp
#if defined(_WIN64)
_INLINE_VAR constexpr size_t _FNV_offset_basis = 14695981039346656037ULL;
_INLINE_VAR constexpr size_t _FNV_prime        = 1099511628211ULL;
#else // defined(_WIN64)
_INLINE_VAR constexpr size_t _FNV_offset_basis = 2166136261U;
_INLINE_VAR constexpr size_t _FNV_prime        = 16777619U;
#endif // defined(_WIN64)

_NODISCARD inline size_t _Fnv1a_append_bytes(size_t _Val, const unsigned char* const _First,
    const size_t _Count) noexcept { // accumulate range [_First, _First + _Count) into partial FNV-1a hash _Val
    for (size_t _Idx = 0; _Idx < _Count; ++_Idx) {
        _Val ^= static_cast<size_t>(_First[_Idx]);
        _Val *= _FNV_prime;
    }
    return _Val;
}
```

with the values being hashed not becoming larger, i.e. after this PR we use `hash_combine` to combine hashes instead of re-hashing hashes.

Reply:
> 1. It may still be nice to have a typedef of HashType, instead of hardcoding `size_t` in various places? I don't feel strongly though.

As a purely aesthetic matter, I like `size_t` a lot. If we do want to abstract it away, I would prefer a name like `hash_t` since it is a simple value rather than a class.

Reply:
Sounds fine about `size_t` to stay as it is then. I don't feel strongly.

About performance, I think it would be good to measure it on some large/interesting codebases. Let me know if you need ideas.

Reply:
Alright, got around to make a little [benchmark thingy](https://github.com/dcodeIO/binaryen-bench) running `wasm-opt@build -O3 fixture.wasm`, yielding the following results for builds of the base and head commits of this PR:

```
Builds:
 * 9c84d90712c1183f1d9ab3e5068d902eb7d627ea
 * hash-refactor

Fixtures:
 * binaryen-O0.wasm
 * binaryen-O3.wasm

Using fixture binaryen-O0.wasm ...
 - wasm-opt@9c84d90712c1183f1d9ab3e5068d902eb7d627ea took 3.493s (mean of 18 samples, min 3.369s)
 - wasm-opt@hash-refactor took 3.418s (mean of 18 samples, min 3.307s)

Using fixture binaryen-O3.wasm ...
 - wasm-opt@9c84d90712c1183f1d9ab3e5068d902eb7d627ea took 2.012s (mean of 24 samples, min 1.987s)
 - wasm-opt@hash-refactor took 1.997s (mean of 24 samples, min 1.967s)
```

According to my test, this PR is slightly faster on both, perhaps because `hash_combine` is more efficient than `rehash`ing hashes and there are no unnecessary type conversions anymore. The fixtures are standalone `binaryen_wasm.wasm` builds generated by commenting out `-s SINGLE_FILE` with varying opt levels (made one unoptimized and one pre-optimized to be sure). I ran the tests in WSL, and the repo also has instructions if you'd like to run locally.

Reply:
Sounds good, thanks @dcodeIO !